### PR TITLE
Autobatch

### PR DESCRIPTION
**Pull request by beckdaniel**
This pull request implement a new training regimen that leverages DyNet autobatching. It's very similar to SimpleTrainingRegimen but accumulates losses sequentially before computing the forward and backward passes. `batch_size` is expected to be fixed at 1 while `update_every` will encode the "actual batch size".

This request also add a few YAML files to benchmark standard batching vs. autobatching.

### CODE REVIEW

**Comment by beckdaniel** [comment]
Hmmm, I'm bit of a newbie with Github reviews... Should I just click on "Resolve conversation" on the points I addressed? Do I ask for another review when I addressed all the requested changes?

**Comment by msperber** [comment]
Sure, but you can also just write a message when you're done and I'll take another look, either way is fine.

I think checking that the batch size is set through ```update_every``` and not through the batcher and potentially throwing a ```ValueError``` would be a good idea. 

Regarding the profiling, that would also be interesting information! We could probably also merge without that, but if you're planning on doing some profiling we could wait for that. In this case, you can prepend [WIP] to the title of the PR to communicate that this is not ready to merge.

File: xnmt/train/regimens.py [hunk 1]
```diff
@@ -181,6 +182,161 @@ def update(self, trainer: optimizers.XnmtOptimizer) -> None:
...
+          self.update(self.trainer, total_loss)
+          total_loss_val = total_loss.get_factored_loss_val(comb_method=self.loss_comb_method)
+          reported_trg = batchers.ListBatch(total_trg)
+          self.train_loss_tracker.report(reported_trg, total_loss_val)
+          total_loss = FactoredLossExpr()
+          total_trg = []
+          self.checkpoint_and_save(save_fct)
+        if self.should_stop_training(): break
+
+  def checkpoint_and_save(self, save_fct):
```
[the remark this answers is not in the document]
Comment (beckdaniel) [reply xnmt/train/regimens.py]: Done. There is a little hack to call backward at the right time though, with a comment. Not sure if it is the best solution.

File: xnmt/train/regimens.py [hunk 2]
```diff
@@ -181,6 +182,161 @@ def update(self, trainer: optimizers.XnmtOptimizer) -> None:
...
+      total_trg = []
+      for src, trg in self.next_minibatch():
+        if self.dev_zero:
+          self.checkpoint_and_save(save_fct)
+          self.dev_zero = False
+        with utils.ReportOnException({"src": src, "trg": trg, "graph": utils.print_cg_conditional}):
+          #####
+          #if self.num_updates_skipped == 0:
+          #  dy.renew_cg(immediate_compute=settings.IMMEDIATE_COMPUTE, check_validity=settings.CHECK_VALIDITY)
+          #####
```
[the remark this answers is not in the document]
Comment (beckdaniel) [reply xnmt/train/regimens.py]: Done.

File: xnmt/train/regimens.py [hunk 3]
```diff
@@ -181,6 +182,161 @@ def update(self, trainer: optimizers.XnmtOptimizer) -> None:
       assert 0 < self.num_updates_skipped < self.update_every
 
 
+class AutobatchTrainingRegimen(train_tasks.SimpleTrainingTask, TrainingRegimen, Serializable):
+  """
```
[the remark this answers is not in the document]
Comment (beckdaniel) [reply xnmt/train/regimens.py]: Done.

**Comment by beckdaniel** [comment]
Ok, I think it is ready for another review now. The only that should be missing from the previous review is to unify the YAML files into a single one: not sure exactly how to do that.

As for the profiling, I'm ok with this being merged without profiling since it might be useful as is. I can always do another pull request with an improved version later.

**Comment by msperber** [comment]
Sure, that sounds good! The code looks good to me, and I'll send a PR to simplify the config later.

**Comment by philip30** [comment]
Addressing #543

**Pull request merged**